src/hardware/input/intel8042.cpp

```diff
@@ -371,6 +371,63 @@ static uint8_t get_translated(const uint8_t byte)
 	return translation_table[byte];
 }
 
+// ***************************************************************************
+// Port disable workaround
+// ***************************************************************************
+
+// TODO: This part is a workaround for 'Ultima VIII: Pagan' problem with
+// non-working keyboard. The proper fix would require larger rework:
+// - it seems that even if keyboard is disabled, the 8042 should fire the
+//   interrupt
+// - testing with 86Box revealed that command 0xad (disable keyboard port) and
+//   0xae (enable port) are being sent by BIOS with each and every key press;
+//   our internal BIOS does not work this way
+//
+// Ignoring command 0xad totally might be risky, as Windows 3.11 for Workgroups
+// disables/reenables the keyboard port during startup and shutdown, issuing
+// controller commands in between; if something disturbs it, we might run into
+// a problem with keyboard/mouse not responding at all.
```

Comment:
just some naive comments / questions:

- Does original DOSBox just ignores this (and says _"Yes, it's disabled"_) but really does nothing?

- Are there benefits to really honoring the "shutdown"? Maybe the keyboard timers aren't updated when shutdown? Or some other benefits, like clearing "held" states (like capslock/numlock)?

I like that you're getting closer and closer to how the hardware behaved!

Reply:
> Does original DOSBox just ignores this (and says "Yes, it's disabled") but really does nothing?

Basically, yes. Search for `keyb.active` in their [keyboard.cpp](https://sourceforge.net/p/dosbox/code-0/HEAD/tree/dosbox/trunk/src/hardware/keyboard.cpp) file.

> Are there benefits to really honoring the shutdown? Maybe the keyboard timers aren't updated when shutdown? Or clearing "held" states (like capslock/numlock)?

This is not exactly a shutdown, it does not directly affect the microcontroller in the keyboard - it just prevents the scancodes from being sent from the keyboard to the 8042 PS/2 controller; they are not even discarded, they are waiting in a queue inside the keyboard.
This way the guest software can issue commands to the mouse or to the controller without worrying that suddenly a key scancode appears on the input. Windows tends to do this during startup/shutdown.
If we ignore the disable request, there is a very small (but real) risk, that key pressed/released in the wrong moment disrupts whatever the guest software is doing.

Reply:
> This way the guest software can issue commands to the mouse or to the controller without worrying that suddenly a key scancode appears on the input. Windows tends to do this during startup/shutdown.

Thanks for that explanation, @FeralChild64 - makes sense!